certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,560 @@
+    calldataarg args;
+    f(e, args);
+
+    address pendingRootAfter = getPendingRoot();
+
+    // if the function f changed the root, then f must be setPendingRoot or claimRoot.
+    assert pendingRootBefore != pendingRootAfter =>
+        ( f.selector == sig:setPendingRoot(address).selector ||
+        f.selector == sig:claimRoot().selector ),
+        "Pending root changed by a function other than setPendingRoot or claimRoot.";
+    // if the function changed the pending root, the sender was pending root or execution helper.
+    assert pendingRootBefore != pendingRootAfter =>
+        e.msg.sender == pendingRootBefore ||
+        e.msg.sender == executionHelper,
+        "Pending root changed by somebody, who was not root or executor.";
+    assert (pendingRootBefore == 0 && pendingRootAfter != 0) => e.msg.sender == executionHelper,
+        "Pending root changed from 0 by other entity than the executer";
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be canceled again.
+rule scheduledExecutionCanBeCancelledOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    cancel@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be executed again.
+rule scheduledExecutionCanBeExecutedOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    execute@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// User cannot become an executor for already scheduled execution, meaning that
+// a user has to be supplied as one of the executors when scheduling an execution,
+// to be able to execute this execution.
+rule cannotBecomeExecutorForAlreadyScheduledExecution(env e, method f) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    address user;
+    require(length < max_uint256);
+    require(id < length);
+    bool isExecutor = isExecutor(id, user);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    assert !isExecutor => !isExecutor(id, user),
+        "User became an executor after the scheduled execution has been scheduled.";
+}
+
+
+// STATUS - verified
+// An execution can be executed (`canExecute`) only when it has not yet been cancelled
+// or executed and its `executableAt` is not greater than current timestamp.
+rule whatCanBeExecuted(env e) {
+    uint256 id;
+    uint256 length = getSchedExeLength();
+    require(id < length);
+
+    bool canExecute = canExecute(e, id);
+
+    assert (
+        !getSchedExeExecuted(id) &&
+        !getSchedExeCancelled(id) &&
+        getSchedExeExecutableAt(id) < e.block.timestamp
+    ) => canExecute;
```

Comment:
I removed the word "only", because when I tried to make it equivalence, it [failed](https://vaas-stg.certora.com/output/40577/fd82186154b3409b9b075745af3aa822/?anonymousKey=1aa335bc2391d2222ecdb7b74d22b57a306b83ac) and the rule is not worth investigating why, when it would only copy the functionality of the canExecute function. This rule is here because of historic reasons and it has been created by splitting one bigger rule into two I think.

Reply:
Yes I think it is covered by other rules, so that's fine